code/modules/reagents/chemistry/machinery/chem_dispenser.dm

```diff
@@ -288,73 +287,92 @@
 			replace_beaker(usr)
 			. = TRUE
 		if("dispense_recipe")
-			if(!is_operational || QDELETED(cell))
-				return
-
-			var/list/chemicals_to_dispense = saved_recipes[params["recipe"]]
-			if(!LAZYLEN(chemicals_to_dispense))
+			if(!is_operational)
 				return
-			for(var/key in chemicals_to_dispense)
-				var/reagent = GLOB.name2reagent[translate_legacy_chem_id(key)]
-				var/dispense_amount = chemicals_to_dispense[key]
-				if(!dispensable_reagents.Find(reagent))
-					return
-				if(!recording_recipe)
-					if(!beaker)
-						return
-
-					var/datum/reagents/holder = beaker.reagents
-					var/to_dispense = max(0, min(dispense_amount, holder.maximum_volume - holder.total_volume))
-					if(!to_dispense)
-						continue
-					if(!cell?.use(to_dispense / powerefficiency))
-						say("Not enough energy to complete operation!")
-						return
-					holder.add_reagent(reagent, to_dispense, reagtemp = dispensed_temperature)
-					work_animation()
-				else
-					recording_recipe[key] += dispense_amount
-			. = TRUE
+			var/recipe_to_use = params["recipe"]
+			var/list/chemicals_to_dispense = process_recipe_list(recipe_to_use)
+			for(var/key in chemicals_to_dispense) // i suppose you could edit the list locally before passing it
+				var/list/keysplit = splittext(key," ")
+				var/r_id = GLOB.name2reagent[translate_legacy_chem_id(keysplit[1])]
+				if(beaker && dispensable_reagents.Find(r_id)) // but since we verify we have the reagent, it'll be fine
+					var/datum/reagents/R = beaker.reagents
+					var/free = R.maximum_volume - R.total_volume
+					var/rounded = is_resolution(chemicals_to_dispense[key]) ? chemicals_to_dispense[key] : round(chemicals_to_dispense[key], macroresolution)
+					var/actual = min(rounded, (cell.charge * powerefficiency)*10, free)
+					if(actual)
+						if(!cell.use(abs(actual) / powerefficiency))
+							say("Not enough energy to complete operation!")
+							return
+						if(actual > 0)
+							R.add_reagent(r_id, actual)
+						if(actual < 0)
+							R.remove_reagent(r_id, abs(actual))
+						work_animation()
 		if("clear_recipes")
 			if(!is_operational)
 				return
 			var/yesno = tgui_alert(usr, "Clear all recipes?",, list("Yes","No"))
 			if(yesno == "Yes")
 				saved_recipes = list()
 			. = TRUE
-		if("record_recipe")
+		if("record_recipe") //I'm not renaming this lol
 			if(!is_operational)
 				return
-			recording_recipe = list()
-			. = TRUE
-		if("save_recording")
-			if(!is_operational)
-				return
-			var/name = stripped_input(usr,"Name","What do you want to name this recipe?", "Recipe", MAX_NAME_LEN)
+			var/name = stripped_input(usr,"Name","What do you want to name this recipe?(This is broken)", "Recipe", MAX_NAME_LEN)
```

Comment:
Remove this comment.

Reply:
It's not fucking resolved you dickweed, REMOVE THIS COMMENT

Reply:
clueless